File: code/stuff/github_gbdts/github_chinese1/gbdt.py
'''
Created on Nov 30st, 2019
Tree-Based Regression Methods
@author: Shihao Liu
'''

import numpy as np
import logging 
import argparse

from decision_tree import DecisionTree, Data

parser = argparse.ArgumentParser()
parser.add_argument('-maximum_height', default=3, type=int)
parser.add_argument('-iterations', default=20, type=int)
parser.add_argument('-lr', default=0.5, type=float)

args = parser.parse_args()

logging.basicConfig(level = logging.INFO,format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class GBDT():
    def __init__(self, lr=0.3, iterations=80, maximum_height=10):
        self.forest = []
        self.learning_rate = lr
        self.iterations = iterations
        self.lr_list = []
        self.maximum_height = maximum_height

    def fit(self, dataset):
        base_constant_prediction = np.mean(dataset[:, -1])
        base_tree = DecisionTree(is_leaf=True, split_value=base_constant_prediction, maximum_height=self.maximum_height)
        residuals = dataset[:, -1] - base_constant_prediction
        self.lr_list.append(1)
        self.forest.append(base_tree)
        features = dataset[:, :-1]
        for i in range(self.iterations):
            dataset = np.concatenate((features, residuals[:, None]), axis=1)
            tree = DecisionTree(maximum_height=self.maximum_height)
            tree.build_tree(tree, dataset)
            for i, example in enumerate(dataset):
                residuals[i] -= self.learning_rate * tree.predict(tree, example)
            self.forest.append(tree)
            self.lr_list.append(self.learning_rate)

    def predict(self, dataset):
        loss = 0
        for example in dataset:
            prediction = 0
            for (lr, tree) in zip(self.lr_list, self.forest):
            # for tree in self.forest:
                prediction += lr * tree.predict(tree, example)
            loss += (example[-1] - prediction) ** 2
            logger.info("真实输出为{:.2f}，预测值为{:.2f}".format(example[-1], prediction))
        logger.info("********** 模型的loss为{:.2f} **********".format(loss))

if __name__ == '__main__':
    # 训练
    data = Data('bikeSpeedVsIq_train.txt')
    training_set = data.load_file()
    gbdt = GBDT(lr=args.lr, iterations=args.iterations, maximum_height=args.maximum_height)
    gbdt.fit(training_set)

    # 预测
    data = Data('./bikeSpeedVsIq_test.txt')
    test_set = data.load_file()
    gbdt.predict(test_set)

    # sklearn自带的波士顿房价数据，及其gbdt训练预测。
    # from sklearn.datasets import load_boston
    # from sklearn.model_selection import train_test_split
    # from sklearn.ensemble import GradientBoostingRegressor
    # data = load_boston()
    # X_train, X_test, y_train, y_test = train_test_split(data["data"],data["target"],test_size=0.3, random_state=0)
    # gbmodel = GradientBoostingRegressor()
    # gbmodel.fit(X_train, y_train)
    # gbdt.predict(test_set)

File: code/stuff/github_gbdts/github_chinese1/gbdt.cpp
#include<vector>
#include<iostream>

#include "data.h"
#include "decision_tree.h"

using namespace std;

class GBDT
{
    public:
    vector<shared_ptr<DecisionTree>> forest;
    vector<float> lr_list;
    float learning_rate;
    int iterations;
    int maximum_height;
    GBDT
    (
        float learning_rate = 0.5,
        int iterations = 10,
        int maximum_height = 3
    ): learning_rate(learning_rate), iterations(iterations), maximum_height(maximum_height) {};
    void fit(FeaturesLabels&);
    vector<float> predict(FeaturesLabels&);
};

void GBDT::fit(FeaturesLabels& dataset)
{
    double sum = 0;
    vector<float> residuals;

    // 首先是将所有label的平均值作为base
    for (size_t i = 0; i < dataset.features.size(); ++i)
    {
        sum += dataset.labels[i];
    }
    float base_prediction = sum / dataset.features.size();
    auto base_tree = make_shared<DecisionTree>(1, 5, true, -1, base_prediction);

    for (size_t i = 0; i < dataset.features.size(); ++i)
    {
        residuals.push_back(dataset.labels[i] - base_prediction);
    }
    lr_list.push_back(1);
    forest.push_back(base_tree);

    //构建多棵树形成boosting，与此同时更新残差。
    for (int i = 0; i < iterations; ++i)
    {
        auto tree = make_shared<DecisionTree>(1, 5, false, -1);
        tree -> build_tree(tree, dataset);
        for (size_t j = 0; j < dataset.features.size(); ++j)
        {
            residuals[j] -= learning_rate * tree -> predict(tree, dataset.features[j]);
        }
        dataset.labels = residuals;
        forest.push_back(tree);
        lr_list.push_back(learning_rate);
    }
}

vector<float> GBDT::predict(FeaturesLabels& dataset)
{
    double loss = 0;
    vector<float> predictions;
    for (size_t i = 0; i < dataset.features.size(); ++i)
    {
        float pred = 0;
        for (size_t j = 0; j < forest.size(); ++j)
        {
            pred += (forest[j] -> predict(forest[j], dataset.features[i])) * lr_list[j];
        }
        predictions.push_back(pred);
        loss += (pred - dataset.labels[i]) * (pred - dataset.labels[i]);
        cout << dataset.labels[i] << " === " << predictions[i] << endl;
    }
    cout << "Total loss is " << loss;
    return predictions;
}

int main()
{
    Data* data = LoadData("bikeSpeedVsIq_train.txt");
    FeaturesLabels training_set = split_features_labels(data);
    GBDT gbdt_model(0.1, 50, 3); // 参数对应为lr，iterations，maximum height
    gbdt_model.fit(training_set);

    /* data所指内存已经在split_features_labels函数中被delete，但data本身仍存在 */
    data = LoadData("bikeSpeedVsIq_test.txt"); 
    FeaturesLabels test_set = split_features_labels(data);
    gbdt_model.predict(test_set);

    return 0;
}